### PR TITLE
test(lints): add lint rw::await-in-loop

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Inspired by #13970 

Found several bugs, also many false positive cases.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by TennyZhuang** [comment]
image
image

**Comment by TennyZhuang** [comment]
I'll try to eliminate false positive cases in the further PRs.

Also I'll fix the true positive cases to improve the performance.

**Review by BugenZhao — COMMENTED** [review]
Would you mind sharing some results found by this lint?

File: lints/src/await_in_loop.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+impl<'hir, 'tcx> Visitor<'hir> for AwaitInLoopVisitor<'hir, 'tcx> {
+    fn visit_expr(&mut self, ex: &'hir Expr<'_>) {
+        match &ex.kind {
+            ExprKind::Loop(_block, _label, source, _span)
+                // In most cases, a raw `loop` can't be concurrent.
+                if !matches!(source, LoopSource::Loop) =>
+            {
+                if !is_from_async_await(ex.span) {
```
Comment (xxchan) [inline lints/src/await_in_loop.rs:89]: What does this `if` do? :eyes:

**Review by xxchan — COMMENTED** [review]
sgtm

**Comment by TennyZhuang** [comment]
[await_in_loop.log](https://github.com/risingwavelabs/risingwave/files/13701348/await_in_loop.log)

**Comment by TennyZhuang** [comment]
```
warning: `risingwave_error` (lib) generated 1 warning
warning: `risingwave_udf` (lib) generated 1 warning
warning: `risingwave_rpc_client` (lib) generated 3 warnings
warning: `risingwave_expr` (lib) generated 3 warnings
warning: `risingwave_hummock_trace` (lib) generated 4 warnings
warning: `risingwave_storage` (lib) generated 49 warnings
warning: `risingwave_jni_core` (lib) generated 1 warning
warning: `risingwave_connector` (lib) generated 26 warnings (5 duplicates)
warning: `risingwave_meta` (lib) generated 90 warnings
warning: `risingwave_batch` (lib) generated 11 warnings
warning: `pgwire` (lib) generated 1 warning
warning: `risingwave_frontend` (lib) generated 13 warnings
warning: `risingwave_expr_impl` (lib) generated 4 warnings
warning: `risingwave_meta_service` (lib) generated 1 warning
warning: `risingwave_compactor` (lib) generated 1 warning
warning: `risingwave_stream` (lib) generated 39 warnings
warning: `risingwave_compute` (lib) generated 2 warnings
warning: `risingwave_ctl` (lib) generated 9 warnings
warning: `risingwave_e2e_extended_mode_test` (lib) generated 5 warnings
warning: `risingwave_compaction_test` (lib) generated 7 warnings
warning: `risingwave_regress_test` (lib) generated 3 warnings
warning: `risingwave_bench` (bin "s3-bench") generated 3 warnings
warning: `risingwave_state_cleaning_test` (bin "risingwave_state_cleaning_test") generated 3 warnings
```

**Pull request closed without merging**